File: combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        size = len(candidates)
        res = []
        
        def getCombinations(idx,sum_,temp):
            if sum_ == target:
                res.append(temp)
                return
            if idx>=size:
                return
            
            for i in range(idx,size):
                if sum_+candidates[i]<=target:
                    getCombinations(i,sum_+candidates[i],temp+[candidates[i]])
            return
        
        getCombinations(0,0,[])
        return res

File: combination-sum/combination-sum.java
class Solution {
    
    public static void recurSum(int[] candidates, ArrayList<Integer> currArr, List<List<Integer>> combinations, int idx, int target, int sum){
        
        if(sum == target){
            if(combinations.contains(currArr) == false){
                combinations.add((ArrayList<Integer>)currArr.clone());   
            }
            return;
        }
        else if(idx>=candidates.length){
            return;
        }
        
        if((sum + candidates[idx])<=target){
            currArr.add(candidates[idx]);
            recurSum(candidates,currArr,combinations,idx,target,sum+candidates[idx]);
            recurSum(candidates,currArr,combinations,idx+1,target,sum+candidates[idx]);
            currArr.remove(currArr.size()-1);
            
        }
        recurSum(candidates,currArr,combinations,idx+1,target,sum);
        
        return;
        
    }
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        
        List<List<Integer>> resCombinations = new ArrayList<List<Integer>>();
        recurSum(candidates,new ArrayList<Integer>(),resCombinations,0,target,0);
        return resCombinations;
    }
}